Add Source PornPics

marioplus opened the pull request:
Closes: #745 

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

marioplus committed:
feat(pornpics): add source pornpics

marioplus committed:
feat(pornpics): add i18n support

marioplus committed:
fix(pornpics): Default category read error

marioplus committed:
fix(pornpics): Content-Type is not set correctly

marioplus committed:
fix(pornpics): properly handle searches

marioplus committed:
fix(pornpics): properly handle searches

marioplus committed:
fix(pornpics): correct next page detection logic.

Add +1 to requested image count per page,Compare actual received count with pageSize to determine next page.

marioplus committed:
fix(pornpics): set base language to en

marioplus committed:
fix(pornpics): safely handle gallery info parsing

marioplus committed:
feat(pornpics): add filter

marioplus committed:
chore(pornpics): remove unused dependency

marioplus committed:
fix(pornpics): correct category urlPart values.

marioplus committed:
refactor(pornpics): simplify category browsing logic

- Remove category unselected
- Treat all non-search requests as category browsing

AwkwardPeak7 commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/JsonFileLoader.kt:
```diff
@@ -0,0 +1,26 @@
+import kotlinx.serialization.json.decodeFromStream
+import uy.kohesive.injekt.injectLazy
+
+class JsonFileLoader {
```
make a singelton object and remove comanion object

```kt
object JsonFileLoader {
```

AwkwardPeak7 commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPics.kt:
```diff
@@ -0,0 +1,200 @@
+    override fun latestUpdatesRequest(page: Int) = buildMangasPageRequest(page, 2)
+
+    private fun buildMangasPageRequest(page: Int, period: Int): Request {
+        val categoryOption = PornPicsPreferences.getCategoryOption(getPreferences())
```
put `preference` in class to reuse

AwkwardPeak7 commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPics.kt:
```diff
@@ -0,0 +1,200 @@
+        chapter_number = 0F
+        setUrlWithoutDomain(element.absUrl("href"))
+        name = intl["chapter.name.default"]
+    }
```
not really using the response here, override `fetchChapterList` and return chapter from there without making network call

marioplus committed:
refactor(pornpics): make a singelton object and remove comanion object

marioplus committed:
refactor(pornpics): put in class to reuse preference

marioplus committed:
refactor(pornpics): optimize chapter loading with fetchChapterList

- Replace chapterFromElement with fetchChapterList
- Reduce unnecessary network requests

marioplus committed:
fix(pornpics): correct CategoryType initialization

vetleledaal commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/JsonFileLoader.kt:
```diff
@@ -0,0 +1,20 @@
+
+    inline fun <reified T> loadJsonAs(fileName: String): T {
+        val fileContent = classLoader.getResourceAsStream(fileName)
+        return json.decodeFromStream<T>(fileContent)
```
How does this handle errors? Failing loudly might be valid.

vetleledaal commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPics.kt:
```diff
@@ -0,0 +1,208 @@
+        val builder = when {
+            isDefaultCategory -> "$baseUrl/popular/api/galleries/list"
+            period == 1 -> "$baseUrl/$categoryOption"
+            period == 2 -> "$baseUrl/$categoryOption/recent"
```
Using an int as a sentinel value here is a bad idea in terms of readability. Can you instead make the input an optional `suffix`?
Like:
```kt
override fun popularMangaRequest(page: Int) {
    val categoryOption = PornPicsPreferences.getCategoryOption(preferences)
    buildMangasPageRequest(page, "/$categoryOption")
}

override fun latestUpdatesRequest(page: Int) {
    val categoryOption = PornPicsPreferences.getCategoryOption(preferences)
    buildMangasPageRequest(page, "/$categoryOption/recent")
}
```

marioplus replied:
Use boolean parameters and add parameter source explanation
See commit: 28327576

vetleledaal commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPicsExt.kt:
```diff
@@ -0,0 +1,49 @@
+internal fun HttpUrl.Builder.addQueryParameter(encodedName: String, encodedValue: Int) =
+    addQueryParameter(encodedName, encodedValue.toString())
+
+internal fun HttpUrl.Builder.addQueryPageParameter(page: Int) =
```
So it lines up with `.addQueryParameterPage()`:
```suggestion
internal fun HttpUrl.Builder.addQueryParameterPage(page: Int) =
```

vetleledaal commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPics.kt:
```diff
@@ -0,0 +1,208 @@
+    ): Request {
+        val categoryOption = activeCategoryTypeSelector.selectedCategoryOption(filters)
+        val sortUrlPart = filters.firstInstance<PornPicsFilters.SortSelector>().toUriPart()
+        val isSearch = categoryOption.toUrlPart().contains('?')
```
Seems brittle in case they change it in the future. Resolve on ACK.

marioplus replied:
Extract individual methods and make the judgment logic more robust
See commit: a9f7970d

vetleledaal commented on src/all/pornpics/src/eu/kanade/tachiyomi/extension/all/pornpics/PornPicsExt.kt:
```diff
@@ -0,0 +1,49 @@
+    urlPart: String?,
+    addPath: Boolean = true,
+    addQuery: Boolean = true,
+): HttpUrl.Builder {
```
Can you not access the original `HttpUrl` object from a builder? Should prefer their parsing for query parameters.

marioplus committed:
refactor(pornpics): improve method naming

- Rename `addQueryPageParameter` to `addQueryParameterPage` for clarity

marioplus committed:
refactor(pornpics): improve API readability with boolean parameters

- Change `buildMangasPageRequest(page: Int, period: Int)` to:
  `buildMangasPageRequest(page: Int, popular: Boolean)`
- Replace numeric period flag with semantic boolean
- Simplify request building logic

marioplus committed:
refactor(pornpics): extract category search logic to dedicated method

- Extract `useSearch` as standalone method
- Add enhanced validation logic for category search

marioplus committed:
refactor(pornpics): replace manual parsing with HttpUrl

- Replace custom URL parsing logic with HttpUrl utility

marioplus committed:
fix(pornpics): remove invalid category options

marioplus committed:
refactor(pornpics): improve JSON error handling

- Throw specific exception type when JSON parsing fails
